### PR TITLE
Gravity simulation

### PR DESCRIPTION
**Pull request by ParthPan7**
1. Added Gravity simulation 

**Note** : Merge UI_Menu pull request prior merging Gravity for maintaining change sequences.

### CODE REVIEW

File: PhysicsExploration/GravitySimulation.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#include "GravitySimulation.h"
+
+float GravitySimulation::u;
+float GravitySimulation::v;
+float GravitySimulation::s0;
+float GravitySimulation::s;
+Gravity* GravitySimulation::gravity;
```
Comment (RoseDuf) [inline PhysicsExploration/GravitySimulation.cpp:7]: add space under here to separate function

File: PhysicsExploration/Gravity.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	v = u + g * t
+
+	s = next position of centre of mass an object
+	s = s0 + u * t + 1/2 * g * t^2
+*/
+class Gravity
+{
+private:
+	const float gravitationalConstant = 9.807;
+	const float marsGravitationalConstant = 3.721;
```
Comment (RoseDuf) [inline PhysicsExploration/Gravity.h]: use capital letters for constants 
```
const float EARTH_GRAVITATIONAL_ACCELERATION = 9.807;
const float MARS_GRAVITATIONAL_ACCELERATION = 3.721;
```

File: PhysicsExploration/GravitySimulation.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+	shader.setBool("hasDirLight", _hasDirLight);
+	shader.setBool("hasPointLight", _hasPointLight);
+	shader.setBool("hasSpotLight", _hasSpotLight);
+}
+
+void GravitySimulation::updatePhysicsParameters()
+{
+	v = gravity->calculateVelocity(u, _deltaTime);
+	s = gravity->calculatePositionFromVelocity(s0, u, _deltaTime);
```
Comment (RoseDuf) [inline PhysicsExploration/GravitySimulation.cpp:63]: do variables v and s need to be declared as class variables? couldn't they be local to this function?

File: PhysicsExploration/GravitySimulation.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+GravitySimulation::~GravitySimulation()
+{
+}
+
+void GravitySimulation::renderScene(const Shader& shader) const
+{
+	shader.setVector("spotLight.position", glm::vec4(_camera->GetPosition(), 1.0f));
+	shader.setVector("spotLight.direction", _camera->GetFront());
+	shader.setVector("view_position", _camera->GetPosition());
+	updatePhysicsParameters();
```
Comment (RoseDuf) [inline PhysicsExploration/GravitySimulation.cpp:41]: I don't see the new position s0 being applied anywhere? am I missing something?

Comment (ParthPan7) [reply PhysicsExploration/GravitySimulation.cpp:41]: Actually s is being used if you check code properly instead s0. s0 is there to keep track of last s. I can use s0 making s local.

Comment (ParthPan7) [reply PhysicsExploration/GravitySimulation.cpp:41]: Actually, declaring operands while in frame is taking too much of time to allocate memory. I remember why I declare them all global Rose :) It's halting animation.

Comment (RoseDuf) [reply PhysicsExploration/GravitySimulation.cpp:41]: interesting!

Comment (RoseDuf) [reply PhysicsExploration/GravitySimulation.cpp:41]: but that still didn't answer my question. Where is the new position s (or s0) being applied?

Comment (ParthPan7) [reply PhysicsExploration/GravitySimulation.cpp:41]: > but that still didn't answer my question. Where is the new position s (or s0) being applied?

Would you like to search following line in a file ? see it uses s.  Please check **renderScene** function.
model_matrix = glm::translate(model_matrix, glm::vec3(0.0f, -1.0f * s, 0.0f));

Comment (RoseDuf) [reply PhysicsExploration/GravitySimulation.cpp:41]: I didn't see it, thanks! :P

**Pull request merged**